packages/form-core/src/FormApi.ts

```diff
@@ -41,7 +58,15 @@ export type FormValidateAsyncFn<
   value: TFormData
   formApi: FormApi<TFormData, TFormValidator>
   signal: AbortSignal
-}) => ValidationError | Promise<ValidationError>
+}) => FormValidationError<TFormData> | Promise<FormValidationError<TFormData>>
+
+export type FormValidator<TFormData, Type, Fn = unknown> = {
```

Comment:
Small tip that @tannerlinsley shared with me: Prefix all generics with `T` so you don't lose track of what's a generic and what's not.

`Type` would become `TType` and `Fn` -> `TFn`